fix(cupertino-nav-bar): use transparent backgroundColor on largeTitle

cpboyd opened the pull request:
## Description

As described in the #60411, Apple's iOS apps typically don't shade the background until it's in the collapsed state.
To have an all-around solution to match any sort of background on the underlying content, I used `Colors.transparent`.

Collapsed:
![IMG_0532](https://user-images.githubusercontent.com/6323077/86013395-19d50500-b9ed-11ea-9273-6184e216ae01.jpg)

Expanded:
![IMG_0531](https://user-images.githubusercontent.com/6323077/86013397-1a6d9b80-b9ed-11ea-9acd-5ee0a696330c.jpg)

Potentially this could also allow a customizable `largeBackgroundColor` with a default of `CupertinoTheme.of(context).scaffoldBackgroundColor`, but I figured `Colors.transparent` would be easiest for a drop-in fix with minimal impact on existing apps.

```dart
final largeBackgroundColor = CupertinoDynamicColor.resolve(widget.largeBackgroundColor, context)
    ?? CupertinoTheme.of(context).scaffoldBackgroundColor;
...
    final Widget navBar = _wrapWithBackground(
      border: border,
      backgroundColor: showLargeTitle
          ? CupertinoDynamicColor.resolve(largeBackgroundColor, context)
          : CupertinoDynamicColor.resolve(backgroundColor, context),
...
```

## Related Issues

Fixes #60411.

## Tests

I added the following tests:

None yet. Starting this PR to get better discussion about the best path forward for this change.
If a `largeBackgroundColor` field is desired, then obviously different tests are needed.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [x] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

The `Large title golden` test failed, as this has an impact on the `backgroundColor`.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

cpboyd committed:
[cupertino] transparent nav_bar on largeTitle

Piinks reviewed (commented):
Hi @cpboyd welcome! Can you add a test to verify the behavior of your change? Thanks!

cpboyd commented:
@Piinks If simply using transparent as I've done (rather than adding a new parameter), I think it should be sufficient to update the `nav_bar_test.large_title.png`.

I'll have to read up on the golden image testing and try to get that done.

cpboyd committed:
[cupertino] nav_bar_test for largeTitle
transparent default background

skia-gold commented:
Gold has detected about 1 untriaged digest(s) on patchset 2.
View them at https://flutter-gold.skia.org/cl/github/61242

cpboyd commented:
@Piinks I added a widget test to check for the transparent background color and rebased against the current master.

I'm still not completely clear on my responsibilities regarding the existing Flutter Gold test per https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package:flutter

I ran the `flutter test --update-goldens test/cupertino/nav_bar_test.dart`. The diffs seem to be related to the fact that the background is now transparent.

Piinks reviewed (commented):
Hey @cpboyd the image diffs can be viewed at https://flutter-gold.skia.org/search?crs=github&fdiffmax=-1&fref=false&frgbamax=255&frgbamin=0&head=true&include=false&issue=61242&limit=50&master=false&match=name&metric=combined&neg=false&offset=0&pos=false&query=source_type%3Dflutter&sort=desc&unt=true if they are as expected, we can approve them and land your change. :)
@xster or @willlarche can you take a look at this change to match native behavior?

Piinks commented:
The cirrus failure here is due to https://github.com/flutter/flutter/issues/67463, fix is in https://github.com/flutter/flutter/pull/67468
Our migration to luci has caused a lot of gold issues lately, should be resolved shortly. :)

DanielEdrisian commented:
Thanks @cpboyd for pointing this out. I did a check locally on Xcode and verified it's indeed transparent in a large title.

> a drop-in fix with minimal impact on existing apps.

Are we sure about this?

Would you be able to implement largeBackgroundColor and have it default to backgroundColor? Let's give developers control over the largeBackgroundColor as needed. This will prevent a possible break.

cpboyd commented:
@DanielEdrisian I can work on that change

Hixie commented:
@cpboyd is splitting these into several PRs something you'd like to do? I'm sorry we're so slow at dealing with these PRs, I'm trying to figure out the best next step here.

cpboyd commented:
@Hixie Unfortunately, other obligations have distracted me from my flutter projects. I'll close the PR for now. If someone else wants to fork and take up the work, they're free to do so.
